### ISSUE TITLE
Docson schema browser: problems handling anyOf, allOf and complex schemas

### ISSUE THREAD

**Issue by kd-ods** [openownership/data-standard#682]
**Summary of the bug or issue**
We use docson on the [schema-browser](https://standard.openownership.org/en/main/standard/schema-browser.html#schema-browser) page to produce a browsable version of the schema. Now the urn-resolving issues are being fixed, it's clear that there are other issues with how docson is handling jsonschema 2020-12.

I'm logging them here, rather than against docson, because I think that the decisions about what to do need to be made here first. (I'm looking at the schema browser page on a local build of the schema-reference-edits branch, where the urn-resolving issues are fixed.)

Sorry for the formatting... lots of screenshots below.

**Additional context and details**

The following is not an exhaustive list, but gives examples of each type of problem.

1) 'allOf' problems

a) Hanging 'allOf':

<img src="https://github.com/openownership/data-standard/assets/35495305/e15b68e7-c55c-4999-b2f3-d8b3403ab96c)" width=50%>

This happens because the allOf is a set of if-then constraints to match the `recordType` (entity, person, relationship) with the right schema for `recordDetails`. See the [statement.json schema](https://github.com/openownership/data-standard/blob/main/schema/statement.json).

Similar issue can be seen under the Annotations object in the browser.

b) EntityType sub-properties not displayed at all:

<img src="https://github.com/openownership/data-standard/assets/35495305/c3c8d165-b299-4cbd-b486-cfeec6901e52)" width=50%>

This happens because the allOf logic, which tests `type` values align with `subtype` values. See the [entity-record.json schema](https://github.com/openownership/data-standard/blob/main/schema/entity-record.json).

2) 'anyOf' problems

a) Publisher sub-properties not displayed at all:

<img src="https://github.com/openownership/data-standard/assets/35495305/a10a827f-0450-4052-a7f8-f46fafba79b3)" width=50%>

This is in the `publicationDetails` object. The `publisher` object has sub-properties that aren't displayed. The anyOf conditions have different properties being required. See the [statement.json schema](https://github.com/openownership/data-standard/blob/main/schema/statement.json).

b) Hanging 'any of' for the Identifier object:

<img src="https://github.com/openownership/data-standard/assets/35495305/37c7efc0-b3cd-4f93-aef4-bbb3e2e81fef)" width=50%>

See the [components.json schema](https://github.com/openownership/data-standard/blob/104a600399d14c4c2413e5c317f13a5ab6417646/schema/components.json#L90). The anyOf conditions have different properties being required.

3) Overlapping schemas problems

a) Constraint on type of address for Person placeOfBirth field:

<img src="https://github.com/openownership/data-standard/assets/35495305/19724b77-792e-4209-b3eb-2e4ffd0fc794)" width=50%>

This shows up very oddly in the browser. It seems that the single valid enum value means the toggle button has the field name on it. See the [person-record.json schema](https://github.com/openownership/data-standard/blob/104a600399d14c4c2413e5c317f13a5ab6417646/schema/person-record.json#L58) .

Compare to to how multiple valid enums are displayed for the person `addresses` field:

<img src="https://github.com/openownership/data-standard/assets/35495305/13bcd2ef-c6fb-4682-b9e1-70b43e7e5597)" width=50%>

4) Alternative types or formats for a property - UI not great. Examples:

<img src="https://github.com/openownership/data-standard/assets/35495305/8da90e0d-b31a-47c3-a029-ca64b165db86)" width=50%>

and this:

<img src="https://github.com/openownership/data-standard/assets/35495305/0abcc4a3-c68b-4d2e-87d9-54ab5681414c)" width=50%>

and this:

<img src="https://github.com/openownership/data-standard/assets/35495305/8a5bde5a-5ad9-43a7-b8ae-66974847df27)" width=50%>

End.

**Suggested resolution**

I think (3) and (4) - we might just have to leave. But (1) and (2) result in subproperties not actually appearing in the browser.

Expecting docson to handle complex schema logic is a BIG ask imo. Most of the constraints that we implement using complex logic are communicated via property descriptions and therefore do not need to be handled and communicated via the docson UI. For our use case, we could strip out all of the allOf and anyOf clauses (I think!) and we would get around these docson issues and allow the schema to be browsed. (And I suppose that could be done via docson if `ignoreAnyOf` true/false and `ignoreAllOf` true/false options were added.)

What's the developer view on this?

**Comment by kd-ods** [comment]
@radix0000 , @rhiaro - We're going to need your input to a decision on this over the next couple of days.

**Comment by kd-ods** [comment]
> For our use case, we could strip out all of the allOf and anyOf clauses (I think!) and we would get around these docson issues

I should have said: where an object's properties simply weren't displaying (`entityType` and `publisher`), this was fixed by removing the allOf/anyOf logic.

**Comment by kd-ods** [comment]
@radix0000 - when I built and tested the docs with the docson update, everything looked good locally, and we [merged the PR](https://github.com/openownership/data-standard/pull/689). BUT the [build on ReadTheDocs](https://standard.openownership.org/en/main/standard/schema-browser.html) isn't showing the docson widgets. (Rachel's local build also doesn't show the docson widgets.)

What do you think the issue could be?? With the submodule update on my local machine, 'git status' was giving:

```
Your branch is up to date with 'origin/schema-reference-edits'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   ../docs/_static/docson (new commits)

```

Did I need to commit those changes in docs/_static/docson??

**Comment by kd-ods** [comment]
Fixed by #695
